File: src/simulation/physics/vpm/advection.js
/**
 * Forward Euler advection: x_{n+1} = x_n + dt · v(x_n)
 * [STATUS]: CORRECT — O(h) first-order, simple and stable
 */
export function advectParticles(particles, dt) {
  for (let i = 0; i < particles.length; i += 1) {
    const particle = particles[i]
    const vx = particle.flowVx ?? 0
    const vy = particle.flowVy ?? 0
    const vz = particle.flowVz ?? 0

    particle.px = particle.x
    particle.py = particle.y
    particle.pz = particle.z
    particle.x += vx * dt
    particle.y += vy * dt
    particle.z += vz * dt
    particle.vx = vx * dt
    particle.vy = vy * dt
    particle.vz = vz * dt
  }
}

/**
 * Save current positions for RK2 restore.
 */
export function capturePositions(particles) {
  const count = particles.length
  const positions = new Float64Array(count * 3)
  for (let i = 0; i < count; i += 1) {
    positions[i * 3] = particles[i].x
    positions[i * 3 + 1] = particles[i].y
    positions[i * 3 + 2] = particles[i].z
  }
  return positions
}

/**
 * Restore positions from snapshot.
 */
export function restorePositions(particles, positions) {
  const count = Math.min(particles.length, Math.floor(positions.length / 3))
  for (let i = 0; i < count; i += 1) {
    particles[i].x = positions[i * 3]
    particles[i].y = positions[i * 3 + 1]
    particles[i].z = positions[i * 3 + 2]
  }
}

/**
 * Save current flow velocities (v1 for RK2).
 */
export function captureFlowVelocities(particles) {
  const count = particles.length
  const velocities = new Float64Array(count * 3)
  for (let i = 0; i < count; i += 1) {
    velocities[i * 3] = particles[i].flowVx ?? 0
    velocities[i * 3 + 1] = particles[i].flowVy ?? 0
    velocities[i * 3 + 2] = particles[i].flowVz ?? 0
  }
  return velocities
}

/**
 * RK2 (midpoint) final advection: x_{n+1} = x_n + dt · v(x_mid)
 * Called after second velocity evaluation at midpoint positions.
 * [STATUS]: CORRECT — O(h²) second-order
 */
export function advectParticlesRK2Final(particles, dt, savedPositions) {
  const count = Math.min(particles.length, Math.floor(savedPositions.length / 3))
  for (let i = 0; i < count; i += 1) {
    const particle = particles[i]
    const vx = particle.flowVx ?? 0
    const vy = particle.flowVy ?? 0
    const vz = particle.flowVz ?? 0

    const x0 = savedPositions[i * 3]
    const y0 = savedPositions[i * 3 + 1]
    const z0 = savedPositions[i * 3 + 2]

    particle.px = x0
    particle.py = y0
    particle.pz = z0
    particle.x = x0 + vx * dt
    particle.y = y0 + vy * dt
    particle.z = z0 + vz * dt
    particle.vx = vx * dt
    particle.vy = vy * dt
    particle.vz = vz * dt
  }
}

/**
 * Compute CFL-limited time step.
 * dt_cfl = C · h / max|v|, where h ≈ σ (core radius as proxy for spacing).
 */
export function computeCflDt(particles, params) {
  const count = particles.length
  if (count === 0) return { cflDt: Infinity, maxSpeed: 0 }

  let maxSpeed = 0
  for (let i = 0; i < count; i += 1) {
    const vx = particles[i].flowVx ?? 0
    const vy = particles[i].flowVy ?? 0
    const vz = particles[i].flowVz ?? 0
    const speed = Math.sqrt(vx * vx + vy * vy + vz * vz)
    if (speed > maxSpeed) maxSpeed = speed
  }

  if (maxSpeed <= 1e-8) return { cflDt: Infinity, maxSpeed }

  const h = Math.max(params.coreRadiusSigma ?? params.minCoreRadius ?? 0.01, 1e-4)
  const cflSafety = Math.max(params.cflSafety ?? 0.4, 0.05)
  const cflDt = cflSafety * h / maxSpeed

  return { cflDt, maxSpeed }
}

File: torus-core/crates/torus-physics/src/advection.rs
//! Particle advection: Euler and RK2 (midpoint).
//!
//! Port of `src/simulation/physics/vpm/advection.js`.

use glam::DVec3;

use crate::particle::Particle;
use crate::params::SimParams;

/// Forward Euler: x_{n+1} = x_n + dt · v(x_n). O(h).
pub fn advect_euler(particles: &mut [Particle], dt: f64) {
    for p in particles.iter_mut() {
        p.prev_position = p.position;
        p.position += p.flow_velocity * dt;
        p.velocity = p.flow_velocity * dt;
    }
}

/// Capture current positions for RK2 restore.
pub fn capture_positions(particles: &[Particle]) -> Vec<DVec3> {
    particles.iter().map(|p| p.position).collect()
}

/// Restore positions from snapshot.
pub fn restore_positions(particles: &mut [Particle], positions: &[DVec3]) {
    let count = particles.len().min(positions.len());
    for i in 0..count {
        particles[i].position = positions[i];
    }
}

/// RK2 final step: x_{n+1} = x_0 + dt · v(x_mid). O(h²).
///
/// Called after second velocity evaluation at midpoint positions.
pub fn advect_rk2_final(particles: &mut [Particle], dt: f64, saved_positions: &[DVec3]) {
    let count = particles.len().min(saved_positions.len());
    for i in 0..count {
        let x0 = saved_positions[i];
        particles[i].prev_position = x0;
        particles[i].position = x0 + particles[i].flow_velocity * dt;
        particles[i].velocity = particles[i].flow_velocity * dt;
    }
}

/// CFL-limited time step: dt_cfl = C · h / max|v|.
pub fn compute_cfl_dt(particles: &[Particle], params: &SimParams) -> CflResult {
    if particles.is_empty() {
        return CflResult {
            cfl_dt: f64::INFINITY,
            max_speed: 0.0,
        };
    }

    let max_speed = particles
        .iter()
        .map(|p| p.flow_velocity.length())
        .fold(0.0_f64, f64::max);

    if max_speed <= 1e-8 {
        return CflResult {
            cfl_dt: f64::INFINITY,
            max_speed,
        };
    }

    let h = params.core_radius_sigma.max(params.min_core_radius).max(1e-4);
    let cfl_safety = 0.4_f64.max(0.05); // default 0.4
    let cfl_dt = cfl_safety * h / max_speed;

    CflResult { cfl_dt, max_speed }
}

#[derive(Debug, Clone, Copy)]
pub struct CflResult {
    pub cfl_dt: f64,
    pub max_speed: f64,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn euler_advection_moves_particle() {
        let mut particles = vec![Particle {
            flow_velocity: DVec3::new(1.0, 0.0, 0.0),
            ..Particle::new(0, DVec3::ZERO)
        }];

        advect_euler(&mut particles, 0.1);
        assert!((particles[0].position.x - 0.1).abs() < 1e-10);
        assert_eq!(particles[0].prev_position, DVec3::ZERO);
    }

    #[test]
    fn rk2_uses_saved_positions() {
        let saved = vec![DVec3::new(1.0, 2.0, 3.0)];
        let mut particles = vec![Particle {
            flow_velocity: DVec3::new(0.5, 0.0, 0.0),
            ..Particle::new(0, DVec3::new(999.0, 999.0, 999.0))
        }];

        advect_rk2_final(&mut particles, 0.1, &saved);
        assert!((particles[0].position.x - 1.05).abs() < 1e-10);
        assert!((particles[0].position.y - 2.0).abs() < 1e-10);
    }

    #[test]
    fn cfl_dt_finite_for_moving_particles() {
        let params = SimParams {
            core_radius_sigma: 0.1,
            ..SimParams::default()
        };
        let particles = vec![Particle {
            flow_velocity: DVec3::new(10.0, 0.0, 0.0),
            ..Particle::new(0, DVec3::ZERO)
        }];

        let result = compute_cfl_dt(&particles, &params);
        assert!(result.cfl_dt.is_finite());
        assert!(result.cfl_dt > 0.0);
        assert!((result.max_speed - 10.0).abs() < 1e-10);
    }

    #[test]
    fn cfl_dt_infinite_for_stationary() {
        let params = SimParams::default();
        let particles = vec![Particle::new(0, DVec3::ZERO)];

        let result = compute_cfl_dt(&particles, &params);
        assert!(result.cfl_dt.is_infinite());
    }
}
